docs/design/daemonless.md

```diff
@@ -0,0 +1,139 @@
+gRPC API services we support.
+
+It will listen on a unix socket (or tcp socket) for API requests - provided via
+a positional argument, defaulting to `unix:///var/run/bpfd.sock`.
+It will shutdown after a timeout of inactivity - provided by a `--timeout` flag
+defaulting to 5 seconds.
+
+It will support being run as a systemd service, via socket activation, which
+will allow it to be started on demand when a request is made to the unix socket.
+When in this mode it will not create the unix socket itself, but will instead
+use LISTEN_FDS to determine the file descriptor of the unix socket to use.
+
+Usage in local development (or packaged in a container) is still possible
+by running as follows:
+
+```console
+sudo bpfd --timeout=0 unix:///var/run/bpfd.sock
+```
+
+> :note: The bpfd user and group will be deprecated.
+> We will also remove some of the unit-file complexity (i.e directories)
+> and handle this in bpfd itself.
+
+### Command: bpfd load file
+
+As the name suggests, this command will load an eBPF program from a file.
+This was formerly `bpfctl load-from-file`.
+
+### Command: bpfd load image
+
+As the name suggests, this command will load an eBPF program from a container
+image. This was formerly `bpfctl load-from-image`.
+
+### Command: bpfd unload
+
+This command will unload an eBPF program. This was formerly `bpfctl unload`.
+
+### Command: bpfd list
+
+This command will list the eBPF programs that are currently loaded.
+This was formerly `bpfctl list`.
+
+### Command: bpfd pull
+
+This command will pull the bpfd container image from a registry.
+This was formerly `bpfctl pull`.
+
+### Command: bpfd images
+
+This command will list the bpfd container images that are available.
+This command didn't exist, but makes sense to add.
+
+### Command: bpfd version
+
+This command will print the version of bpfd.
+This command didn't exist, but makes sense to add.
+
+### State Management
+
+This is perhaps the most significant change from how bpfd currently works.
+
+Currently bpfd maintains state in-memory about the programs that have been
+loaded (by bpfd, and the kernel).
+Some of this state is flushed to disk, so if bpfd is restarted it can
+reconstruct it.
+
+Flushing to disk and state reconstruction is cumbersome at present and having
+to move all state management out of in-memory stores is a forcing function
+to improve this.
+We will replace the existing state management with [sled], which gives us
+a familiar API to work with while also being fast, reliable and persistent.
+
+[sled]: https://github.com/spacejam/sled
+
+### Metrics and Monitoring
+
+While adding metrics and monitoring is not a goal of this design, it should
+nevertheless be a consideration. In order to provide metrics to Prometheus
+or OpenTelemetry we will require an additional exporter process.
+
+We can either:
+
+1. Use the bpfd socket and retrieve metrics via the gRPC API
+1. Place state access + metrics gathering functions in a library, such that
```

Comment:
```suggestion
2. Place state access + metrics gathering functions in a library, such that
```

Reply:
Markdown numbered lists allow you to re-use `1.` so that the numbering is sequential.
That way when you add another numbered bullet the lists re-adjusts.
I prefer this as I suck at remembering to re-number lists 😆

Reply:
Noice I actually didn't know that!